// @wix/advanced-seo-utils/dist/es/src/private/item-types/static-page-v2/schema-presets/local-business.js
import {
    siteIDs
} from '../../../adapters/utils';
import {
    SCHEMA_TYPES
} from './schema-types';
var schema = {
    '@context': 'https://schema.org/',
    '@type': 'LocalBusiness',
    name: "{{" + siteIDs.BUSINESS_NAME + "}}",
    url: "{{" + siteIDs.SITE_URL + "}}",
    image: "{{" + siteIDs.BUSINESS_SCHEMA_IMAGE + "}}",
    address: {
        '@type': 'PostalAddress',
        addressCountry: "{{" + siteIDs.BUSINESS_LOCATION_COUNTRY + "}}",
        addressLocality: "{{" + siteIDs.BUSINESS_LOCATION_CITY + "}}",
        addressRegion: "{{" + siteIDs.BUSINESS_LOCATION_STATE + "}}",
        postalCode: "{{" + siteIDs.BUSINESS_POSTAL_CODE + "}}",
        streetAddress: "{{" + siteIDs.BUSINESS_LOCATION_STREET_ADDRESS + "}}",
    },
    telephone: "{{" + siteIDs.BUSINESS_PHONE + "}}",
};
var requiredSchemaVariables = [
    siteIDs.BUSINESS_NAME,
    siteIDs.BUSINESS_LOCATION_FORMATTED,
];
export default {
    type: SCHEMA_TYPES.LOCAL_BUSINESS,
    schema: schema,
    requiredSchemaVariables: requiredSchemaVariables,
};
